soh/soh/ActorDB.cpp

```diff
@@ -0,0 +1,597 @@
+    { ACTOR_DOOR_KILLER, "Fake Door" },
+    { ACTOR_BG_SPOT11_OASIS, "Oasis (Desert Colossus)" },
+    { ACTOR_BG_SPOT18_FUTA, "Goron Jar Lid" },
+    { ACTOR_BG_SPOT18_SHUTTER, "Sliding Doors (Goron City)" },
+    { ACTOR_EN_MA3, "Malon (Adult, Lon Lon Ranch)" },
+    { ACTOR_EN_COW, "Cow" },
+    { ACTOR_BG_ICE_TURARA, "Icicles" },
+    { ACTOR_BG_ICE_SHUTTER, "Vertical Ice Bars (Ice Cavern)" },
+    { ACTOR_EN_KAKASI2, "Pierre the Scarecrow Spawn" },
+    { ACTOR_EN_KAKASI3, "Bonooru the Scarecrow" },
+    { ACTOR_OCEFF_WIPE4, "Scarecrow's Song Ocarina Effect" },
+    { ACTOR_EN_EG, "Void-out Trigger (Tower Collapse)" },
+    { ACTOR_BG_MENKURI_NISEKABE, "False Stone Walls (Gerudo Training Grounds)" },
+    { ACTOR_EN_ZO, "Zora" },
+    { ACTOR_OBJ_MAKEKINSUTA, "Skulltula Sprouting from Bean Spot" },
+    { ACTOR_EN_GE3, "Gerudo Fortress Leader" },
+    { ACTOR_OBJ_TIMEBLOCK, "Time Block" },
+    { ACTOR_OBJ_HAMISHI, "Bronze Boulder" },
+    { ACTOR_EN_ZL4, "Zelda (Child)" },
+    { ACTOR_EN_MM2, "Running Man (Adult Era)" },
+    { ACTOR_BG_JYA_BLOCK, "Silver Block (Child Era)" },
+    { ACTOR_OBJ_WARP2BLOCK, "Navi Infospot (Green, Time Block)" }
+};
+
+ActorDB::ActorDB() {
+    db.reserve(ACTOR_NUMBER_MAX); // reserve size for all initial entries so we don't do it for each
+    for (const AddPair& pair : initialActorTable) {
+        Entry& entry = Add(pair.name, actorDescriptions[pair.init.id], pair.init);
+    }
+}
+
+// Adds an actor at the given index. The name must be unique.
+ActorDB::Entry& ActorDB::Add(const std::string& name, const std::string& desc, size_t index) {
+    assert(!nameTable.contains(name)); // TODO this should maybe throw instead. We'll need to think about error handling for mods that try to declare the same actor.
+
+    if (db.size() < (index + 1)) {
+        db.resize(index + 1);
+    }
+    Entry& newEntry = db.at(index);
+    newEntry.entry.id = index;
+
+    assert(!newEntry.entry.valid);
+
+    nextFreeId = std::max(nextFreeId, index + 1);
+
+    nameTable[name] = newEntry.entry.id;
+    newEntry.setName(name);
+    newEntry.setDesc(desc);
+
+    newEntry.entry.valid = true;
+
+    return newEntry;
+}
+
+// Adds an actor with the legacy ActorInit struct. The id is provided in it.
+ActorDB::Entry& ActorDB::Add(const std::string& name, const std::string& desc, const ActorInit& init) {
+    Entry& entry = Add(name, desc, init.id);
+
+    entry.entry.category = init.category;
+    entry.entry.flags = init.flags;
+    entry.entry.objectId = init.objectId;
+    entry.entry.instanceSize = init.instanceSize;
+    entry.entry.init = init.init;
+    entry.entry.destroy = init.destroy;
+    entry.entry.update = init.update;
+    entry.entry.draw = init.draw;
+    entry.entry.reset = init.reset;
+
+    return entry;
+}
+
+// Adds an actor with the new ActorDBInit struct. The id assigned to the actor is dynamic. Use the return Entry or RetrieveId to get it.
+ActorDB::Entry& ActorDB::Add(const ActorDBInit& init) {
```

Comment:
For consistency I think I'd prefer:
```suggestion
ActorDB::Entry& ActorDB::AddEntry(const ActorDBInit& init) {
```